### ISSUE TITLE
[Feature request] Ternary operator

### ISSUE THREAD

**Issue by Rangi42** [gbdev/rgbds#621]
This would be mostly useful for within user-defined functions (#201), but would also be more terse than the current `if/else/endc`.

Example:

```
giveitem: MACRO
	db giveitem_command
	db \1 ; item
if _NARG == 2
	db \2 ; quantity
else
	db 1 ; quantity
endc
ENDM

; simplifies to

giveitem: MACRO
	db giveitem_command
	db \1 ; item
	db _NARG == 2 ? \2 : 1 ; quantity
ENDM
```

```
growth_rate: MACRO
; [1]/[2]*n**3 + [3]*n**2 + [4]*n - [5]
	dn \1, \2
	if \3 & $80 ; signed
		db -\3 | $80
	else
		db \3
	endc
	db \4, \5
ENDM

; simplifies to

growth_rate: MACRO
; [1]/[2]*n**3 + [3]*n**2 + [4]*n - [5]
	dn \1, \2
	db \3 & $80 ? -\3 | $80 : \3 ; signed
	db \4, \5
ENDM
```

**Comment by ISSOtm** [comment]
Might as well make it short-circuiting too (#619).

**Comment by Rangi42** [comment]
Lazy evaluation would make this more useful, but it's worth adding even before then. For example, this:

```
growth_rate: MACRO
; [1]/[2]*n**3 + [3]*n**2 + [4]*n - [5]
	dn \1, \2
	if \3 < 0 ; two's complement -> signed magnitude
		db -\3 | $80
	else
		db \3
	endc
	db \4, \5
ENDM
```

would become:

```
growth_rate: MACRO
; [1]/[2]*n**3 + [3]*n**2 + [4]*n - [5]
	dn \1, \2
	db \3 < 0 ? -\3 | $80 : \3 ; two's complement -> signed magnitude
	db \4, \5
ENDM
```

**Comment by Rangi42** [comment]
Also note that to be consistent with other operators, including mere parentheses, this should turn strings into numbers. Which would disallow things like `PRINTT \1 > 1 ? "items" : "item"`.

**Comment by ISSOtm** [comment]
Well, RGBASM is actually very strongly typed, with expression syntax depending on the expected type, so the ternary operator could be separately implemented for strings.

**Comment by Rangi42** [comment]
True, but should it? And if so/regardless, should `T_LPAREN relocexpr T_RPAREN` be split into two cases for numbers and strings? (It feels wrong to me that just parenthesizing an expression should change its meaning.) AFAIK all the other cases (`+x`, `-x`, `x & y`, `HIGH(x)`, etc) are fine with the coercion to number with `str2int2`.

**Comment by aaaaaa123456789** [comment]
> Well, RGBASM is actually very strongly typed, with expression syntax depending on the expected type, so the ternary operator could be separately implemented for strings.

Is it? `db "a"` is ambiguous.

**Comment by Rangi42** [comment]
`db`, `dw`, and `dl` are special-cased for strings (see parser.y).

**Comment by Rangi42** [comment]
The [Pasmo](http://pasmo.speccy.org/pasmodoc.html#operators) assembler also has a `?:` ternary operator; it and the `&&` and `||` logical operators are short-circuited.

**Comment by Rangi42** [comment]
Another use case for ternary operators without waiting for lazy evaluation (even if they end up both being added in the same release): [pokecrystal16](https://github.com/aaaaaa123456789/pokecrystal16/blob/expand-move-ID/engine/16/macros.asm#L156-L161)'s code-generating macros (a bad idea for small patterns, but necessary for cases like this).

```
		srl h
		; if HIGH(\1) is even, srl h is correct; otherwise, we have to invert it. Since we can't conditionally include a
		; ccf instruction (as it would create a circular dependency between assembler and linker), we have no choice but
		; to execute a ccf on odd HIGH(\1) and a nop on even HIGH(\1) by inserting the correct instruction dynamically
		; using a db. Note that ccf = $3F and nop = $00.
		db (HIGH(\1) & 1) * $3F
```

This would be clearer as:

```
		srl h
		db HIGH(\1) & 1 ? $3F : $00
```

**Comment by ISSOtm** [comment]
I think `? :` shouldn't be a problem... hopefully.
